1) Error. Foo is in the function and not outside of it. 

2) prints bar.  Line 1 passes to line 7. While line 6 is called its not outputted.

4) 
function name: multiply_numbers
function arguments: 2, 3, 4
function definition: def - everything on lines 1-3
function body: result = num1 * num2 * num3 - everything on lines 2 and 3
function parameters: num1, num2, num3
function invocation:  multiply_numbers(2, 3, 4)
function return value: 24
all identifiers: multiply_numbers, num1, num2, num3, result, and product

5) Nothing is printed.

6) Nothing is printed. Print statement is inside the function and doesn't get used.

7) Error message: Expected 2 arguments, 1 was returned.

8) Error message: Expected 2 arguments, 3 passed.

9.) Prints: (42, 3.141592, 2.718). The defaults weren't needed and thus not passed.

10.) Prints: 42, 3.141592, 2 

11.) Prints 42, 3, 2

12) Error, nothing was passed to the first and there's no first default, so the code ends.

13) Error, when nothing gets passed to the 3rd, when a 3rd default is required because the 2nd default was innacted.

14) Everything but the def, =, f 

15) float, print, return  = built in
left, right prompt, = local
first, second, product = global

16)
multiply: line 1,9
(left, right) line 1, line 2
def get_num(prompt), 4 7 8
(prompt) 4, 5
float 5
input 5
print 5

17) function names: def say(message)
    built in: print, input, max
    method: upper, lower